src/boot/efi/cpio.c

```diff
@@ -0,0 +1,465 @@
+                        continue;
+
+                d = StrDuplicate(dirent->FileName);
+                if (!d)
+                        return log_oom();
+
+                if (n_items+2 > n_allocated) {
+                        UINTN m;
+
+                        if (n_items > (UINTN_MAX / sizeof(UINT16)) - 16) /* Overflow check, just in case */
+                                return log_oom();
+
+                        m = n_items + 16;
+                        items = ReallocatePool(items, n_allocated * sizeof(UINT16*), m * sizeof(UINT16*));
+                        if (!items)
+                                return log_oom();
+
+                        n_allocated = m;
+                }
+
+                items[n_items++] = TAKE_PTR(d);
+                items[n_items] = NULL; /* Let's always NUL terminate, to make freeing via strv_free() easy */
+        }
+
+        if (n_items == 0) {
+                /* Empty directory */
+                *ret_buffer = NULL;
+                *ret_buffer_size = 0;
+                return EFI_SUCCESS;
+        }
+
+        /* Now, sort the files we found, to make this uniform and stable (and to ensure the TPM measurements
+         * are not dependent on read order) */
+        sort_pointer_array((VOID**) items, n_items, (compare_pointer_func_t) StrCmp);
+
+        /* Generate the leading directory inodes right before adding the first files, to the
+         * archive. Otherwise the cpio archive cannot be unpacked, since the leading dirs won't exist. */
+        err = pack_cpio_prefix(target_dir_prefix, dir_mode, &inode, &buffer, &buffer_size);
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed to pack cpio prefix: %r", err);
+
+        for (UINTN i = 0; i < n_items; i++) {
+                _cleanup_freepool_ CHAR8 *content = NULL;
+                UINTN contentsize;
+
+                err = file_read(extra_dir, items[i], 0, 0, &content, &contentsize);
+                if (EFI_ERROR(err)) {
+                        log_error_status_stall(err, L"Failed to read %s, ignoring: %r", items[i], err);
+                        continue;
+                }
+
+                err = pack_cpio_one(
+                                items[i],
+                                content, contentsize,
+                                target_dir_prefix,
+                                access_mode,
+                                &inode,
+                                &buffer, &buffer_size);
+                if (EFI_ERROR(err))
+                        return log_error_status_stall(err, L"Failed to pack cpio file %s: %r", dirent->FileName, err);
+        }
+
+        err = pack_cpio_trailer(&buffer, &buffer_size);
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed to pack cpio trailer: %r");
+
+#if ENABLE_TPM
+        err = tpm_log_event(
+                        tpm_pcr,
+                        (EFI_PHYSICAL_ADDRESS) (UINTN) buffer,
```

Comment:
(UINTN) shouldn't be needed here. buffer is a pointer, so it should already be the same size as UINTN, right?

Reply:
nope. I thought so too. But it's not the case. EFI_PHYSICAL_ADDRESS is alwas 64bit in UEFI, even on 32bit archs (to deal with PAE I guess). UINTN is defined by uefi to be pointer size. gcc will complain if you cast a pointer to an integer that doesn't match the pointer size. Hence we have to do this in two steps: first make an integer of it, and then enlarge the integer.

Only way to figure this out is by trying to build this stuff on 32bit. I only grokked this once I built this on the 32bit CI.